TauAnalysis/MCEmbeddingTools/plugins/MuMuForEmbeddingSelector.cc

```diff
@@ -61,60 +73,126 @@ class MuMuForEmbeddingSelector : public edm::stream::EDProducer<> {
 //
 // constructors and destructor
 //
-MuMuForEmbeddingSelector::MuMuForEmbeddingSelector(const edm::ParameterSet& iConfig)
+MuMuForEmbeddingSelector::MuMuForEmbeddingSelector(const edm::ParameterSet &iConfig)
     : ZmumuCandidates_(consumes<edm::View<reco::CompositeCandidate>>(
           iConfig.getParameter<edm::InputTag>("ZmumuCandidatesCollection"))) {
-  //register your products
-  /* Examples
-   produces<ExampleData2>();
-
-   //if do put with a label
-   produces<ExampleData2>("label");
- 
-   //if you want to put into the Run
-   produces<ExampleData2,InRun>();
-*/
+  use_zmass = iConfig.getParameter<bool>("use_zmass");
   produces<edm::RefVector<pat::MuonCollection>>();
-
-  //now do what ever other initialization is needed
+  produces<float>("oldMass");
+  produces<float>("newMass");
+  produces<float>("nPairCandidates");
+  produces<bool>("isMediumLeadingMuon");
+  produces<bool>("isTightLeadingMuon");
+  produces<bool>("isMediumTrailingMuon");
+  produces<bool>("isTightTrailingMuon");
+  produces<float>("initialMETEt");
+  produces<float>("initialMETphi");
+  produces<float>("initialPuppiMETEt");
+  produces<float>("initialPuppiMETphi");
+  theVertexLabel_ = consumes<reco::VertexCollection>(iConfig.getParameter<edm::InputTag>("inputTagVertex"));
+  theBeamSpotLabel_ = consumes<reco::BeamSpot>(iConfig.getParameter<edm::InputTag>("inputTagBeamSpot"));
+  theMETLabel_ = consumes<edm::View<pat::MET>>(iConfig.getParameter<edm::InputTag>("Met"));
+  thePuppiMETLabel_ = consumes<edm::View<pat::MET>>(iConfig.getParameter<edm::InputTag>("PuppiMet"));
+  // now do what ever other initialization is needed
 }
 
 //
 // member functions
 //
 
 // ------------ method called to produce the data  ------------
-void MuMuForEmbeddingSelector::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+void MuMuForEmbeddingSelector::produce(edm::Event &iEvent, const edm::EventSetup &iSetup) {
   using namespace edm;
   edm::Handle<edm::View<reco::CompositeCandidate>> ZmumuCandidatesHandle;
   iEvent.getByToken(ZmumuCandidates_, ZmumuCandidatesHandle);
   edm::View<reco::CompositeCandidate> ZmumuCandidates = *ZmumuCandidatesHandle;
-
-  const reco::CompositeCandidate* chosenZCand = nullptr;
+  const reco::CompositeCandidate *chosenZCand = nullptr;
+  const reco::CompositeCandidate *chosenZCand_zmass = nullptr;
+  const reco::CompositeCandidate *chosenZCand_largest = nullptr;
   double massDifference = -1.0;
+  edm::Handle<reco::BeamSpot> beamSpot;
+  iEvent.getByToken(theBeamSpotLabel_, beamSpot);
+  edm::Handle<reco::VertexCollection> vertex;
+  iEvent.getByToken(theVertexLabel_, vertex);
+  edm::Handle<edm::View<pat::MET>> met;
+  iEvent.getByToken(theMETLabel_, met);
+  edm::Handle<edm::View<pat::MET>> puppimet;
+  iEvent.getByToken(thePuppiMETLabel_, puppimet);
+  // get primary vertex
+  reco::Vertex::Point posVtx;
+  reco::Vertex::Error errVtx;
+  std::vector<reco::Vertex>::const_iterator vertexIt = vertex->begin();
```

Comment:
can you use a range-based for loop here?

Reply:
Done in a8eae9f.